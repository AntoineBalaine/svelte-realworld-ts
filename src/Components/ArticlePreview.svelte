<script lang="ts">
	import { enhance } from "$app/forms";
	import type * as ApiTypes from "$lib/ApiTypes";
	export let article: ApiTypes.components["schemas"]["Article"];
	const { author, updatedAt, tagList, title, description, slug } = article;
	let { favoritesCount, favorited } = article;
</script>

<div class="article-preview">
	<div class="article-meta">
		<a href="/profile/@{author?.username}"><img src={author.image} alt="author thumbnail" /></a>
		<div class="info">
			<a href="/profile/@{author?.username}" data-testid="author-username" class="author"
				>{author.username}</a
			>
			<span data-testid="author-updatedAt" class="date">{updatedAt}</span>
		</div>
		<form
			method="POST"
			action="/article/{article.slug}?/toggleFavorite"
			use:enhance={({ form }) => {
				favorited = !favorited;
				favorited ? (favoritesCount += 1) : (favoritesCount -= 1);
				const button = form.querySelector("button");
				if (button) {
					button.disabled = true;
				}
				return (actionresult) => {
					// un-disable the toggle
					if (button) {
						button.disabled = false;
					}
					const { result } = actionresult;
					if (result.type !== "success") {
						// if the form submission fails, reset the form to previous state
						favorited = !favorited;
						favorited ? (favoritesCount += 1) : (favoritesCount -= 1);
					}
				};
			}}
		>
			<input hidden type="checkbox" id="checkbox" name="favorited" bind:value={favorited} />
			<button class="btn btn-outline-primary btn-sm pull-xs-right">
				<i class="ion-heart" />
				{favoritesCount}
			</button>
		</form>
	</div>
	<a href="/article/{slug}" class="preview-link">
		<h1 data-testid="article-title">{title}</h1>
		<p data-testid="article-description">{description}</p>
		<span>Read more...</span>
		<ul class="tag-list">
			{#each tagList as tag}
				<li>
					<a href="/?tag={tag}" class="tag-pill tag-default tag-outline">{tag}</a>
				</li>
			{/each}
		</ul>
	</a>
</div>
