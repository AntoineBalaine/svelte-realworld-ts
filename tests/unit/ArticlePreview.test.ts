import { render } from "@testing-library/svelte";
import { describe, test, expect } from "vitest";
import ArticlePreview from "src/components/ArticlePreview.svelte";

import type * as ApiTypes from "$lib/ApiTypes";

const articleSample: ApiTypes.components["schemas"]["Article"] = {
	slug: "thething",
	title: "The thing",
	description: "this article is about a thing",
	body: "This is the article’s content. Fascinating, isn’t it?",
	tagList: ["fascinating", "article", "thing", "things"],
	/** Format: date-time */
	createdAt: new Date().toString(),
	/** Format: date-time */
	updatedAt: new Date().toString(),
	favorited: false,
	favoritesCount: 55,
	author: {
		username: "el svelte autor",
		bio: "the mysterious svelte component creator",
		image: "",
		following: false
	}
};

describe("VerticalTabs Component", () => {
	test("should render the component", () => {
		const articlePreview = render(ArticlePreview, {
			article: articleSample
		});

		expect(articlePreview.getByTestId("author-username").innerHTML).toEqual(
			articleSample.author.username
		);
		expect(articlePreview.getByTestId("article-title").innerHTML).toEqual(articleSample.title);
		expect(articlePreview.getByTestId("article-description").innerHTML).toEqual(
			articleSample.description
		);
		articlePreview
			.getAllByRole("listitem")
			.forEach((element, index) =>
				expect(element.getElementsByTagName("a")[0].innerHTML).toEqual(articleSample.tagList[index])
			);

		/**
		 * At each render, the test-runner will create a new date,
		 * which will be different from the date displayed on the page.
		 * To avoid this, it’s necessary to swap out the page’s date.
		 */
		articlePreview.getByTestId("author-updatedAt").innerHTML = "place-holder date";

		expect(articlePreview.container).toMatchSnapshot();
	});
});
