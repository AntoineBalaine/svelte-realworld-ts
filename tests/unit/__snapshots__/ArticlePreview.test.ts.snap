// Vitest Snapshot v1

exports[`ArticlePreview Component > should render the component 1`] = `
<body>
  <div>
    <div
      class="article-preview"
    >
      <div
        class="article-meta"
      >
        <a
          href="/profile/@el svelte autor"
        >
          <img
            alt="author thumbnail"
            src=""
          />
        </a>
         
        <div
          class="info"
        >
          <a
            class="author"
            data-testid="author-username"
            href="/profile/@el svelte autor"
          >
            el svelte autor
          </a>
           
          <span
            class="date"
            data-testid="author-updatedAt"
          >
            place-holder date
          </span>
        </div>
         
        <form
          action="/article/thething?/toggleFavorite"
          method="POST"
        >
          <input
            hidden=""
            id="checkbox"
            name="favorited"
            type="checkbox"
            value="false"
          />
           
          <button
            class="btn btn-outline-primary btn-sm pull-xs-right"
          >
            <i
              class="ion-heart"
            />
             
            55
          </button>
        </form>
      </div>
       
      <a
        class="preview-link"
        href="/article/thething"
      >
        <h1
          data-testid="article-title"
        >
          The thing
        </h1>
         
        <p
          data-testid="article-description"
        >
          this article is about a thing
        </p>
         
        <span>
          Read more...
        </span>
         
        <ul
          class="tag-list"
        >
          <li>
            <a
              class="tag-pill tag-default tag-outline"
              href="/?tag=fascinating"
            >
              fascinating
            </a>
             
          </li>
          <li>
            <a
              class="tag-pill tag-default tag-outline"
              href="/?tag=article"
            >
              article
            </a>
             
          </li>
          <li>
            <a
              class="tag-pill tag-default tag-outline"
              href="/?tag=thing"
            >
              thing
            </a>
             
          </li>
          <li>
            <a
              class="tag-pill tag-default tag-outline"
              href="/?tag=things"
            >
              things
            </a>
             
          </li>
        </ul>
      </a>
    </div>
    <!--&lt;ArticlePreview&gt;-->
  </div>
</body>
`;
